brain/circuit.cpp

```diff
@@ -28,6 +28,277 @@
 
 namespace brain
 {
+namespace
+{
+URIs _getMorphologyURIs(const Circuit::Impl& circuit, const GIDSet& gids)
+{
+    const Strings& names = circuit.getMorphologyNames(gids);
+
+    URIs uris;
+    uris.reserve(names.size());
+    for (const auto& name : names)
+        uris.push_back(circuit.getMorphologyURI(name));
+    return uris;
+}
+
+Matrix4fs _getTransforms(const Circuit::Impl& circuit, const GIDSet& gids)
+{
+    const Vector3fs& positions = circuit.getPositions(gids);
+    const Quaternionfs& rotations = circuit.getRotations(gids);
+    if (positions.size() != rotations.size())
+        throw std::runtime_error(
+            "Positions not equal rotations for given GIDs");
+
+    Matrix4fs transforms(positions.size());
+
+#pragma omp parallel for
+    for (size_t i = 0; i < positions.size(); ++i)
+        transforms[i] = Matrix4f(rotations[i], positions[i]);
+    return transforms;
+}
+
+class MorphologyLoader
+{
+public:
+    MorphologyLoader(const Circuit::Impl& impl)
+        : _impl(impl)
+        , _cache(_impl.getMorphologyCache())
+    {
+    }
+
+    neuron::Morphologies load(const GIDSet& gids,
+                              const Circuit::Coordinates coords)
+    {
+        _initRecentering(gids);
+
+        const bool transform = coords == Circuit::Coordinates::global;
+
+        URIs uris = _getMorphologyURIs(_impl, gids);
+        for (auto& uri : uris)
+        {
+            // opt: don't stat abs file path
+            if (uri.getScheme() == "file" && uri.getPath()[0] != '/')
+                uri.setPath(fs::canonical(uri.getPath()).generic_string());
+        }
+
+        // The keys include the information about whether the morphology
+        // needs recentering or not. This is only the case for local
+        // morphologies.
+        const Strings keys = _getKeys(gids, uris, coords);
+        CachedMorphologies cached = _loadFromCache(keys);
+
+        struct MorphologyUse
+        {
+            brion::MorphologyPtr original;
+            brion::MorphologyPtr recentered;
+            size_t numRecentered = 0;
+            size_t numOriginal = 0;
+        };
+        using Loading = std::unordered_map<std::string, MorphologyUse>;
```

Comment:
Why use a type alias?

Reply:
In the beginning I thought I was going to use this type more than once.